apps/emqx_resource/src/emqx_resource.erl

```diff
@@ -418,24 +418,39 @@ query(ResId, Request, Opts) ->
             emqx_resource_buffer_worker:async_query(ResId, Request, Opts)
     end.
 
-get_query_mode_error(ResId, Opts) ->
-    case maps:get(query_mode_cache_override, Opts, true) of
-        false ->
-            case Opts of
-                #{query_mode := QueryMode} ->
-                    {QueryMode, ok};
-                _ ->
-                    {async, unhealthy_target}
-            end;
-        true ->
-            case emqx_resource_manager:lookup_cached(ResId) of
-                {ok, _Group, #{query_mode := QM, error := Error}} ->
-                    {QM, Error};
-                {error, not_found} ->
-                    {error, not_found}
-            end
+get_query_mode_error(_ResId, Request, Opts = #{connector_resource_id := ResId}) ->
+    do_get_query_mode_error(ResId, Request, Opts);
+get_query_mode_error(ResId, Request, Opts) ->
+    do_get_query_mode_error(ResId, Request, Opts).
+
+do_get_query_mode_error(ResId, Request, Opts) ->
+    case emqx_resource_manager:lookup_cached(ResId) of
+        {ok, _Group, ResourceData} ->
+            QM = get_query_mode(ResourceData, Opts),
+            Error = get_error(Request, ResourceData),
+            {QM, Error};
+        {error, not_found} ->
+            {error, not_found}
     end.
 
+get_query_mode(_ResourceData, #{query_mode := QM}) ->
+    QM;
+get_query_mode(#{query_mode := QM}, _Opts) ->
+    QM.
+
+get_error(Request, #{added_channels := #{} = Channels} = ResourceData) when
+    is_tuple(Request), is_binary(element(1, Request))
+->
+    ChannelId = element(1, Request),
+    case Channels of
+        #{ChannelId := #{error := Error}} ->
+            Error;
+        _ ->
+            maps:get(error, ResourceData, undefined)
+    end;
+get_error(_Request, #{error := Error}) ->
+    Error.
```

Comment:
to me, it's really clueless what's going on here.

Reply:
Maybe we should have an `emqx_resource_manager:is_unhealthy_target(ResourceData)` function to try to keep the knowledge about the `ResourceData` structure within `emqx_resouce_manager`. This will make `emqx_resource` simpler at least and the complex logic will be closer to where the structure of the data is "documented".